### PR TITLE
Improve current develop branch

### LINKED ISSUE #31: Mobile app: Improve current develop branch
## Issue Overview

- There has been many PRs that got merged into develop branch, need to re-conciliate them and after that push to `main` branch

## Proposed Solutions or Ideas

- Clean, integrate and test whole app-flow/game-play.
  - Test on multiple device sizes (android, android tablet 7 inch, android tablet 10 inch, iPad 13 inch, iPhone 6inch9)
- Check missing relevant edge cases.
- Propose potential improvements.
- Add script tests.
- Update app screenshots (for listing)

### Ludo
- Fix last player join not displayed bug
- Fix dice roll error handling and animation

### Checkers
- review create game flow and implement missing steps

### Overall
- Adaptive sizing for constant shape widgets (i.e. buttons created from svg elements).
- Review and fix refresh token logic (update the token expiry)
- Try fix the snackbar shown behind dialog 
- Fix login email address input formatting (don't allow spaces) & referral code (auto capitalise)

### PR DESCRIPTION
**Pull request by Emrys02**
Closes #31 

## Issues to work on
- [x] Prevent invalid characters to be entered in auth fields
- [x] Force capitalization on entry of Referral Code
- [x] Checkers flow
- [x] Update `Chevron Border` for buttons
- [x] Submit button
- [x] Contrast on error dialog message
- [x] Prevent exiting game menu while a dialog is visible
- [x] Fix Game Menu bug after leaving a game _ludo game_
- [x] Use Dice to notify players to rolled numbers
- [x] Setup automatic requests for auth tokens
- [x] Provide loading overlay
- [x] Improve Join and Find game flows
- [x] Fix last player joined not displayed
- [x] Fix snackbar shown behind dialog
- [x] Adaptive sizing for constant shape widgets (i.e. buttons created from svg elements).
- [x] Provide form validation on auth email.

## Tests Added
- [x] Snackbar Service.
- [x] AppState.
- [x] UserState.

### CODE REVIEW

**Comment by Emrys02** [comment]
Used a combination of **Input Restriction** _to prevent entry of invalid characters_ and **Input Interception** _to convert input into desired format_

**Comment by Emrys02** [comment]
The commit `Checkers Logic Overhaul` contains the following changes;

* `Game` has been renamed to `GameController`. e.g _LudoGame --> LudoGameController_.
* **Abstract class `MarquisGameController`**: It allows an easy-to-maintain approach to creating game controllers.
* Checkers Main Screen has been added to handle rendering based on PlayState.
* Checkers Flow now uses `PlayState` to render certain screen. e.g _Welcome Screen, Waiting Screen_.
* Instances where `Navigator` is being used to switch between game screen in Checkers Flow have been migrated to update `PlayState` instead.
* `Waiting Room Screen` has been separated from `Create Game Screen` in `Checkers Flow`.
* Support for the Checkers Game `CheckersGameAppPath` has been provided in `AppRouterDelegate`
* Commonly used enums are now available globally.
* Content of Checkers Game Screen has been trimmed, only containing necessary widgets.

**Comment by Emrys02** [comment]
Old ludo waiting room has been removed

**Comment by Emrys02** [comment]
Issue with refresh token was that a request is made a the point of expiry and without support for retries. 

Fix was to start making a request 5mins before token expiry with up to 3 retries performed at 5s intervals. 

A loop is used for the tries rather than a recursion to reduce overhead cost

**Comment by Emrys02** [comment]
All requests with response status codes in the `200s` are now processed as successful requests.

All instances of `showDialog` have their `useRootNavigator` parameter set to true to allow dialogs to be closed indirectly.

**Comment by Emrys02** [comment]
Email Validation Preview

[Checkers Create Game on 11" tablet](https://www.loom.com/share/131f9cf54874456ab018639022d98e98)

[Ludo Create Game on 5" device](https://www.loom.com/share/673aa9fbbcdd4ce3b32199cbd182f7ef)

**Comment by Emrys02** [comment]
`Dependency Injection` was introduced to `AppState` and `User` for use of `HiveBox` and `HttpClient` classes in order to make them testable

**Comment by gianalarcon** [comment]
Looks good to me, just few comments
- For checkers, on min 0:19 we expected to see the manual input box, it was not in the figma, but I remember we implemented manual input box right? @wms2537 @co01y4p ?
- For ludo looks good, I would like to see a longer video, where players roll the dice at least a couple of rounds

**Comment by gianalarcon** [comment]
> Looks good to me, just few comments
> 
> * For checkers, on min 0:19 we expected to see the manual input box, it was not in the figma, but I remember we implemented manual input box right? @wms2537 @co01y4p ?
> * For ludo looks good, I would like to see a longer video, where players roll the dice at least a couple of rounds

reference:  [figma](https://www.figma.com/design/vwxsBUQOK2nKsmzUEDs4xF/Public-Marquis?node-id=35-3733&t=kwE08pzlwYANGdYX-0)

**Comment by Emrys02** [comment]
Manual Token Input

[Game Play](https://www.loom.com/share/30a561608018425599f1a7f5bfab5373?sid=7127e97a-9466-40dc-a592-651eac7e3841) _recommended to watch at 2.5x speed_

**Comment by gianalarcon** [comment]
looks good to me!

File: lib/providers/app_state.dart [hunk 1]
```diff
@@ -2,94 +2,99 @@ import 'dart:async';
...
+
+  AppState({Client? httpClient, Box<AppStateData>? hiveBox})
+      : _httpClient = httpClient,
+        _hiveBox = hiveBox ?? Hive.box<AppStateData>("appState");
 
   @override
   AppStateData build() {
-    _hiveBox ??= Hive.box<AppStateData>("appState");
-    return _hiveBox!.get("appState", defaultValue: AppStateData())!;
+    _httpClient ??= Client();
```
Comment (wms2537) [inline lib/providers/app_state.dart:30]: Set the autoLoginResult to null initially
```
return _hiveBox
        .get("appState", defaultValue: AppStateData())!
        .copyWith(autoLoginResult: null);
```

Comment (wms2537) [reply lib/providers/app_state.dart:30]: So that the app will attempt to refresh token on start.

Current behaviour is that the token is not refreshed on init. So token errors might occur.

File: lib/main.dart [hunk 2]
```diff
@@ -12,6 +12,8 @@ import 'package:marquis_v2/providers/user.dart';
 import 'package:marquis_v2/router/route_information_parser.dart';
 import 'package:marquis_v2/router/router_delegate.dart';
 
+import 'services/snackbar_service.dart';
+
 void main() async {
```
Comment (wms2537) [inline lib/main.dart:17]: Since we modified the structure of the appstate now is different from `main`, better for us to add a check before opening the box for compatibility 
```
try {
    await Hive.openBox<AppStateData>("appState");
  } catch (e) {
    await Hive.deleteBoxFromDisk("appState");
    await Hive.openBox<AppStateData>("appState");
  }
```

Comment (Emrys02) [reply lib/main.dart:17]: I used this suggestion to create a function  `_loadAppStateBox`. Calls to load the hive boxes have been switched to occur asynchronously. 

`await Future.wait([
    _loadAppStateBox(),
    Hive.openBox<UserData>("user"),
    Hive.openBox<LudoSessionData>("ludoSession"),
  ]);`
  
  @wms2537

**Review by wms2537 — APPROVED** [review]
